### ISSUE TITLE
[Feat] Kornia Augmentations tracker

### ISSUE THREAD

**Issue by edgarriba** [kornia/kornia#941]
## More Than Augmentation
In a longer vision, our augmentation module will be more focused on differentiability, with an aim of assisting both training and deployment. We can see there are more researches and applications in this area, with or without Kornia. This module will be the lead. In our next version, we gonna heavily improve:
- the API design and better readability.
- Advanced features like JIT and AMP.
- Better differentiability and more differentiable augmentations.

## 🚀 Feature
Specs for the `kornia.augmentation` module.

- Make it easy to contrib through `AugmentationBase2D`.
- Give support to `torch.jit.script` and `torch.float16`.
- Promote to embed the full code in the main class to keep readability.
- Documentation improvement:
  - Better looking
  - More examples for copying and pasting.
  - Provide a benchmark to notify users that we are better.
- Add `register_buffer` to save static parameters inside ```nn.Module```.

## Motivation

This should be the template to implement a custom augmentation

```python
class RandomHorizontalFlip(AugmentationBase2D):

    def compute_transformation(self, input: torch.Tensor) -> torch.Tensor:
        w: int = input.shape[-1]
        flip_mat: torch.Tensor = torch.tensor([[-1, 0, w - 1],
                                               [0, 1, 0],
                                               [0, 0, 1]], device=input.device, dtype=input.dtype)
        return flip_mat.repeat(input.size(0), 1, 1)

    def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
        return kornia.hflip(input)
```

## User Stories
- _As a user, I want …, so that …_ 
- As a user, I want to decide whether the transforms return the transformation, so that I can use it to chain and potentially apply in sequential.
- As a user, I want to use the augmentations with half precision, so that I can also combine with AMP.
- As a user, I want to be able to run the augmentations with xla devices, so that I can run on TPUs.
- As a user, I want to be able to torchscript the augmentations, so that I can easily serialise and export.
- As a developer, I want to have easy examples of custom augmentations, so that I can contribute with more.

## TODO:
- [x] Refactor engine to deprecate `return_transform` to return only `torch.Tensor` @shijianjian 
  - Transforms will be accessible through class properties
- [ ] Enable torchscript for augmentations https://github.com/kornia/kornia/issues/477
- [x] Remove functional code @shijianjian 
- [x] Fix `RandResizedCrop` critical bug https://github.com/kornia/kornia/issues/927

## List of existing 2d transforms:
- [x] CenterCrop
- [x] ColorJitter
- [x] GaussianBlur
- [x] RandomAffine
- [x] RandomCrop
- [ ] RandomErasing
  - Gradients error. They computed on the non-cropped area than on the mask. Check the gradient image in the bottom. will be fixed in https://github.com/kornia/kornia/pull/873 
- [x] RandomGrayscale
- [x] RandomHorizontalFlip
- [x] RandomVerticalFlip
- [x] RandomMotionBlur
- [x] RandomPerspective
- [x] RandomResizedCrop
  - Broken: https://github.com/kornia/kornia/issues/927
- [x] RandomRotation
- [x] RandomSolarize
- [ ] RandomPosterize
  - Not differentiable (due to the basis of int8 operations). Grad_fn not implemented.
- [x] RandomSharpness
- [ ] RandomEqualize
  - Not backward-propagatable
  - Error message: “derivative for floor_divide is not implemented”
- [x] RandomMixUp
- [x] RandomCutMix

## List of wanted 2d transforms:
- **Adapt from Tormentor:** needs tormentor’s diamond square
  - [x] RandomelasticTransform  @edgarriba
  - [x] RandomThinPlateSpline  @edgarriba
  - [x] RandomFishEye  @edgarriba
    - Adopt from tormentor: https://github.com/anguelos/tormentor#create-custom-augmentation-class
  - [x] RandomPlasmaBrightness
  - [x] RandomPlasmaContrast
  - [ ] RandomPlasmaRgbBrightness
  - [x] RandomPlasmaShadow
  - [ ] ImageCorruption augmentations https://github.com/kornia/kornia/issues/1143
- **Adapt from Albumentations**
  - [x] RandomInvert https://github.com/kornia/kornia/pull/960 @edgarriba 
  - [x] RandomGaussianNoise https://github.com/kornia/kornia/pull/960 @edgarriba 
  - [x] RandomChannelShuffle https://github.com/kornia/kornia/pull/960 @edgarriba 
  - [ ] RandomGlassNoise
  - [x] RandomRGBShift
  - [ ] RandomClahe
  - [x] RandomBoxBlur
  - [ ] RandomMedianBlur
  - [ ] RandomChannelDropout
  - [ ] RandomDownScale
  - [ ] RandomRain -> need `draw_line`
  - [ ] RandomShadow -> need `draw_line`
  - [ ] RandomSnow https://github.com/UjjwalSaxena/Automold--Road-Augmentation-Library/blob/master/Automold.py#L330
  - [ ] RandomSunFlare -> need `draw_line`
- **Containers**
  - [x] VideoSequential
  - [x] OneOf ? or RandomChoice
  - [x] ToTensor https://github.com/kornia/kornia/issues/705
  - [x] Lambda
  - [x] RandomApply
  - [ ] AutoAugment ? (or similar) @shijianjian 
  - [ ] TTA @shijianjian 
  - [ ] Augmentation presets https://github.com/kornia/kornia/issues/1151

Link to discussion docs: https://docs.google.com/document/d/1DcWnoB5L-Ws0asOq9ntXHrCe0ny2NM2WCtLIoU50wQw/edit?usp=sharing

## Advanced features
 - [x] `AugmentationSequential` to support multi-data augmentations
 - [ ] Make `AugmentationSequential` end to end. 
     - [x] Bounding Box #1497 
     - [ ] Mask
     - [x] Keypoints
     - [ ] Label
- [x] Deprecate `return_transform` in favour of `inverse`.
- [ ] AutoAugment pipelines.

-----

## Please, anyone interested: comment below and volunteer yourself :smile:

**Comment by Oleksandra2020** [comment]
Hi, I have a question about why `draw_line` is needed to implement RandomSnow? The Albumentations implementation converts RGB to HLS, increases the lightness and then converts the image back to RGB. Does the use of `draw_line` mean generating random points and drawing white lines in their scope? Thanks in advance.

**Comment by johnnv1** [comment]
@shijianjian Some of the augmentation in the list, are already developed:
- RandomSolarize: https://kornia.readthedocs.io/en/latest/augmentation.module.html#kornia.augmentation.RandomSolarize -- this one differs from a `differentiable solarization`?
- RandomClahe: https://kornia.readthedocs.io/en/latest/augmentation.module.html#kornia.augmentation.RandomClahe
